packages/shopify-app-remix/src/server/authenticate/admin/embedded-authenticate.ts

```diff
@@ -0,0 +1,245 @@
+  ensureValidShopParam,
+  redirectToBouncePage,
+  validateUrlParams,
+} from './helpers/auth-helpers';
+
+interface SessionContext {
+  session: Session;
+  token?: JwtPayload;
+}
+
+const SESSION_TOKEN_PARAM = 'id_token';
+
+export class EmbeddedAuthStrategy<
+  Config extends AppConfigArg,
+  Resources extends ShopifyRestResources = ShopifyRestResources,
+> {
+  protected api: Shopify;
+  protected config: AppConfig;
+  protected logger: Shopify['logger'];
+
+  public constructor({api, config, logger}: BasicParams) {
+    this.api = api;
+    this.config = config;
+    this.logger = logger;
+  }
+
+  public async authenticateAdmin(
+    request: Request,
+  ): Promise<AdminContext<Config, Resources>> {
+    const {api, logger, config} = this;
+    logger.info('shopify-app-remix is local');
+
+    rejectBotRequest({api, logger, config}, request);
+    respondToOptionsRequest({api, logger, config}, request);
+
+    const cors = ensureCORSHeadersFactory({api, logger, config}, request);
+
+    let sessionContext: SessionContext;
+    try {
+      sessionContext = await this.authenticateAndGetSessionContext(request);
+    } catch (errorOrResponse) {
+      if (errorOrResponse instanceof Response) {
+        cors(errorOrResponse);
+      }
+
+      throw errorOrResponse;
+    }
+
+    return createApiContext<Config, Resources>(
+      {api, logger, config},
+      request,
+      sessionContext,
+      cors,
+      handleEmbeddedClientErrorFactory,
+    );
+  }
+
+  private async authenticateAndGetSessionContext(
+    request: Request,
+  ): Promise<SessionContext> {
+    const {api, logger, config} = this;
+    const params: BasicParams = {api, logger, config};
+
+    const url = new URL(request.url);
+
+    const isPatchSessionToken =
+      url.pathname === config.auth.patchSessionTokenPath;
+    const isExitIframe = url.pathname === config.auth.exitIframePath;
+    const isAuthRequest = url.pathname === config.auth.path;
+
+    if (isPatchSessionToken) {
+      logger.debug('Rendering bounce page');
+      throw renderAppBridge(params, request);
+    } else if (isExitIframe) {
+      const destination = url.searchParams.get('exitIframe')!;
+
+      logger.debug('Rendering exit iframe page', {destination});
+      throw renderAppBridge(params, request, {url: destination});
+    } else if (isAuthRequest) {
+      const shop = ensureValidShopParam(request, {api, logger, config});
+      const installUrl = `https://${shop}/admin/oauth/install?client_id=${config.apiKey}`;
+      throw redirect(installUrl);
+    }
+
+    const sessionTokenHeader = getSessionTokenHeader(request);
+    const sessionTokenParam = url.searchParams.get(SESSION_TOKEN_PARAM)!;
+    const sessionTokenString = sessionTokenHeader || sessionTokenParam;
+
+    logger.info('Authenticating admin request');
+
+    if (sessionTokenParam) {
+      await validateUrlParams(request, {api, logger, config});
+      await ensureAppIsEmbeddedIfRequired(request, {api, logger, config});
```

Comment:
This is interesting.

In auth code flow `ensureAppIsEmbeddedIfRequired` makes sense.  In token exchange `IfRequired` makes less sense, because it's always required.

When we product-ionize this code these are the kind of helpers that I'd want to be careful about.  I'd prefer to duplicate the helper and reduce coupling between authe code flow and token exchange